### PR TITLE
Use cuco::static_set in JSON tree algorithm 

### PR DESCRIPTION
**Pull request by karthikeyann**
## Description
In JSON tree algorithms of JSON reader, cuco static_map is used as a set. This PR replaces it with static_set.
No tests are changed. No significant runtime changes.
Addresses part of  #12261

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by karthikeyann** [comment]
Benchmark Comparison:
['before.json', 'after.json']
# nested_json_gpu_parser

## [0] Quadro GV100

|  string_size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|---------------|------------|-------------|------------|-------------|--------------|---------|----------|
|     2^20      |   4.880 ms |       4.49% |   4.898 ms |       5.69% |    18.864 us |   0.39% |   PASS   |
|     2^21      |   4.895 ms |       9.06% |   4.913 ms |       4.03% |    17.902 us |   0.37% |   PASS   |
|     2^22      |   5.317 ms |      10.41% |   5.321 ms |      12.66% |     3.876 us |   0.07% |   PASS   |
|     2^23      |   6.422 ms |       0.50% |   6.412 ms |       0.34% |   -10.405 us |  -0.16% |   PASS   |
|     2^24      |   9.930 ms |      14.53% |   9.787 ms |      14.77% |  -143.676 us |  -1.45% |   PASS   |
|     2^25      |  15.002 ms |      11.11% |  15.006 ms |      11.65% |     3.727 us |   0.02% |   PASS   |
|     2^26      |  22.675 ms |       6.45% |  22.183 ms |       9.47% |  -492.229 us |  -2.17% |   PASS   |
|     2^27      |  40.288 ms |       4.58% |  40.130 ms |       3.07% |  -158.362 us |  -0.39% |   PASS   |
|     2^28      |  76.542 ms |       1.90% |  75.338 ms |       2.95% | -1204.252 us |  -1.57% |   PASS   |
|     2^29      | 145.540 ms |       2.20% | 141.689 ms |       1.44% | -3851.651 us |  -2.65% |   FAIL   |
|     2^30      | 280.676 ms |       1.14% | 273.563 ms |       0.40% | -7112.382 us |  -2.53% |   FAIL   |

# nested_json_gpu_parser_depth

## [0] Quadro GV100

|  depth  |  string_size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|---------|---------------|------------|-------------|------------|-------------|---------------|---------|----------|
|   2^1   |     2^20      |   4.794 ms |       3.81% |   4.844 ms |      13.07% |     50.668 us |   1.06% |   PASS   |
|   2^2   |     2^20      |   4.806 ms |      13.63% |   4.821 ms |       0.32% |     15.013 us |   0.31% |   PASS   |
|   2^3   |     2^20      |   9.694 ms |       0.28% |   9.728 ms |       0.26% |     33.102 us |   0.34% |   FAIL   |
|   2^4   |     2^20      |  11.520 ms |       0.38% |  11.549 ms |       0.32% |     29.233 us |   0.25% |   PASS   |
|   2^1   |     2^22      |   5.921 ms |       0.33% |   5.944 ms |       0.37% |     23.099 us |   0.39% |   FAIL   |
|   2^2   |     2^22      |   5.932 ms |       0.39% |   5.930 ms |       0.64% |     -1.492 us |  -0.03% |   PASS   |
|   2^3   |     2^22      |  10.719 ms |       0.27% |  10.749 ms |       0.29% |     30.260 us |   0.28% |   FAIL   |
|   2^4   |     2^22      |  12.800 ms |       0.27% |  12.836 ms |       0.23% |     36.276 us |   0.28% |   FAIL   |
|   2^1   |     2^24      |  11.261 ms |      10.01% |  11.321 ms |      10.55% |     59.724 us |   0.53% |   PASS   |
|   2^2   |     2^24      |  11.306 ms |      10.39% |  11.296 ms |      10.66% |     -9.278 us |  -0.08% |   PASS   |
|   2^3   |     2^24      |  15.109 ms |       0.90% |  15.232 ms |       4.67% |    122.651 us |   0.81% |   PASS   |
|   2^4   |     2^24      |  18.559 ms |       3.69% |  18.820 ms |       5.51% |    260.069 us |   1.40% |   PASS   |
|   2^1   |     2^26      |  30.345 ms |       4.05% |  30.058 ms |       3.64% |   -287.165 us |  -0.95% |   PASS   |
|   2^2   |     2^26      |  30.467 ms |       4.22% |  30.053 ms |       3.81% |   -413.853 us |  -1.36% |   PASS   |
|   2^3   |     2^26      |  35.087 ms |       2.81% |  35.038 ms |       3.47% |    -49.738 us |  -0.14% |   PASS   |
|   2^4   |     2^26      |  44.072 ms |       3.11% |  43.698 ms |       3.43% |   -373.460 us |  -0.85% |   PASS   |
|   2^1   |     2^28      | 105.235 ms |       2.31% | 103.592 ms |       1.38% |  -1643.400 us |  -1.56% |   FAIL   |
|   2^2   |     2^28      | 105.108 ms |       1.54% | 103.375 ms |       1.34% |  -1732.831 us |  -1.65% |   FAIL   |
|   2^3   |     2^28      | 131.532 ms |       1.10% | 126.514 ms |       1.18% |  -5017.671 us |  -3.81% |   FAIL   |
|   2^4   |     2^28      | 163.348 ms |       1.29% | 160.052 ms |       1.21% |  -3296.169 us |  -2.02% |   FAIL   |
|   2^1   |     2^30      | 402.572 ms |       0.74% | 396.564 ms |       0.50% |  -6008.528 us |  -1.49% |   FAIL   |
|   2^2   |     2^30      | 403.850 ms |       0.71% | 396.296 ms |       0.32% |  -7553.944 us |  -1.87% |   FAIL   |
|   2^3   |     2^30      | 511.063 ms |       0.19% | 489.892 ms |       0.49% | -21171.784 us |  -4.14% |   FAIL   |
|   2^4   |     2^30      | 631.286 ms |       0.30% | 610.291 ms |       0.44% | -20995.106 us |  -3.33% |   FAIL   |

# json_read_data_type

## [0] Quadro GV100

|  data_type  |      io       |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|-------------|---------------|------------|-------------|------------|-------------|---------------|---------|----------|
|    FLOAT    | DEVICE_BUFFER | 769.726 ms |       0.21% | 741.506 ms |       0.07% | -28219.922 us |  -3.67% |   FAIL   |
|   DECIMAL   | DEVICE_BUFFER | 848.174 ms |       0.16% | 826.489 ms |       0.14% | -21684.851 us |  -2.56% |   FAIL   |
|   STRING    | DEVICE_BUFFER | 334.214 ms |       0.48% | 321.141 ms |       0.36% | -13072.504 us |  -3.91% |   FAIL   |
|    LIST     | DEVICE_BUFFER | 246.600 ms |       0.79% | 247.035 ms |       0.76% |    435.057 us |   0.18% |   PASS   |
|   STRUCT    | DEVICE_BUFFER | 936.482 ms |       0.25% | 905.342 ms |       0.35% | -31139.343 us |  -3.33% |   FAIL   |

# json_read_io

## [0] Quadro GV100

|      io       |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|---------------|------------|-------------|------------|-------------|--------------|---------|----------|
|   FILEPATH    | 549.848 ms |       1.97% | 554.247 ms |       1.72% |     4.398 ms |   0.80% |   PASS   |
|  HOST_BUFFER  | 427.906 ms |       0.05% | 426.857 ms |       0.11% | -1049.323 us |  -0.25% |   FAIL   |
| DEVICE_BUFFER | 278.015 ms |       0.50% | 280.157 ms |       0.21% |     2.142 ms |   0.77% |   FAIL   |

# Summary

- Total Matches: 43
  - Pass    (diff <= min_noise): 23
  - Unknown (infinite noise):    0
  - Failure (diff > min_noise):  20

**Comment by karthikeyann** [comment]
> The memory usage should change as a result of this, right? (By not having to store both keys and values)

It makes no difference in peak memory usage, because this part of code does not take the highest memory usage.
But in that particular section of code, it may have reduced usage.

File: cpp/src/io/json/json_tree.cu [hunk 1]
```diff
@@ -36,6 +36,7 @@
 #include <cub/device/device_radix_sort.cuh>
 
 #include <cuco/static_map.cuh>
```
Comment (PointKernel) [inline cpp/src/io/json/json_tree.cu:38]:
```suggestion
```

Comment (karthikeyann) [reply cpp/src/io/json/json_tree.cu:38]: There is another place in this file where `static_map` is used.

Comment (PointKernel) [reply cpp/src/io/json/json_tree.cu:38]: Ah OK. Resovling this comment.

File: cpp/src/io/json/json_tree.cu [hunk 2]
```diff
@@ -632,23 +621,27 @@ std::pair<rmm::device_uvector<size_type>, rmm::device_uvector<size_type>> hash_n
...
-    [d_hashed_cache,
-     d_equal,
-     view       = key_map.get_device_mutable_view(),
-     uq_node_id = col_id.begin()] __device__(auto node_id) mutable {
-      auto it = view.insert_and_find(cuco::make_pair(node_id, node_id), d_hashed_cache, d_equal);
-      uq_node_id[node_id] = (it.first)->first.load(cuda::std::memory_order_relaxed);
-      return it.second;
-    });
+  auto nodes_itr          = thrust::make_counting_iterator<size_type>(0);
+  auto const num_inserted = key_set.insert(nodes_itr, nodes_itr + num_nodes, stream.value());
```
Comment (PointKernel) [inline cpp/src/io/json/json_tree.cu:638]: Can this be `num_columns` directly?
```suggestion
  auto const num_columns = key_set.insert(nodes_itr, nodes_itr + num_nodes, stream.value());
```

File: cpp/src/io/json/json_tree.cu [hunk 3]
```diff
@@ -36,6 +36,7 @@
 #include <cub/device/device_radix_sort.cuh>
 
 #include <cuco/static_map.cuh>
+#include <cuco/static_set.cuh>
```
Comment (ttnghia) [inline cpp/src/io/json/json_tree.cu:39]: Still need static map? If not, please remove the header above.

File: cpp/src/io/json/json_tree.cu [hunk 4]
```diff
@@ -632,23 +621,27 @@ std::pair<rmm::device_uvector<size_type>, rmm::device_uvector<size_type>> hash_n
...
 
+  constexpr size_type empty_node_index_sentinel = -1;
+  using hash_table_allocator_type = rmm::mr::stream_allocator_adaptor<default_allocator<char>>;
+  using hasher_type               = decltype(d_hashed_cache);
+
+  auto key_set = cuco::experimental::static_set{
+    cuco::experimental::extent{compute_hash_table_size(num_nodes)},
+    cuco::empty_key<cudf::size_type>{empty_node_index_sentinel},
+    d_equal,
+    cuco::experimental::linear_probing<1, hasher_type>{d_hashed_cache},
```
Comment (ttnghia) [inline cpp/src/io/json/json_tree.cu:632]: Why using `linear_probing` here? Do we have other options?

**Comment by karthikeyann** [comment]
After replacing static_map with static_set in "node type + field name hashing". 
['before.json', 'after2.json']
# nested_json_gpu_parser

## [0] Quadro GV100

|  string_size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|---------------|------------|-------------|------------|-------------|---------------|---------|----------|
|     2^20      |   4.880 ms |       4.49% |   4.611 ms |       0.37% |   -268.940 us |  -5.51% |   FAIL   |
|     2^21      |   4.895 ms |       9.06% |   4.774 ms |       0.32% |   -120.961 us |  -2.47% |   FAIL   |
|     2^22      |   5.317 ms |      10.41% |   5.190 ms |      11.42% |   -126.517 us |  -2.38% |   PASS   |
|     2^23      |   6.422 ms |       0.50% |   6.204 ms |       0.38% |   -217.953 us |  -3.39% |   FAIL   |
|     2^24      |   9.930 ms |      14.53% |   9.615 ms |      18.52% |   -315.873 us |  -3.18% |   PASS   |
|     2^25      |  15.002 ms |      11.11% |  13.463 ms |      11.93% |  -1538.816 us | -10.26% |   PASS   |
|     2^26      |  22.675 ms |       6.45% |  20.569 ms |       5.89% |  -2105.785 us |  -9.29% |   FAIL   |
|     2^27      |  40.288 ms |       4.58% |  37.852 ms |       5.66% |  -2436.356 us |  -6.05% |   FAIL   |
|     2^28      |  76.542 ms |       1.90% |  67.596 ms |       2.47% |  -8946.637 us | -11.69% |   FAIL   |
|     2^29      | 145.540 ms |       2.20% | 129.453 ms |       0.18% | -16087.117 us | -11.05% |   FAIL   |
|     2^30      | 280.676 ms |       1.14% | 254.863 ms |       0.24% | -25813.133 us |  -9.20% |   FAIL   |

# nested_json_gpu_parser_depth

## [0] Quadro GV100

|  depth  |  string_size  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|---------|---------------|------------|-------------|------------|-------------|---------------|---------|----------|
|   2^1   |     2^20      |   4.794 ms |       3.81% |   4.758 ms |       1.03% |    -35.237 us |  -0.74% |   PASS   |
|   2^2   |     2^20      |   4.806 ms |      13.63% |   4.759 ms |       0.33% |    -47.073 us |  -0.98% |   FAIL   |
|   2^3   |     2^20      |   9.694 ms |       0.28% |   9.639 ms |       0.50% |    -55.950 us |  -0.58% |   FAIL   |
|   2^4   |     2^20      |  11.520 ms |       0.38% |  11.431 ms |       0.21% |    -89.160 us |  -0.77% |   FAIL   |
|   2^1   |     2^22      |   5.921 ms |       0.33% |   5.803 ms |       0.35% |   -117.992 us |  -1.99% |   FAIL   |
|   2^2   |     2^22      |   5.932 ms |       0.39% |   5.807 ms |       0.33% |   -124.324 us |  -2.10% |   FAIL   |
|   2^3   |     2^22      |  10.719 ms |       0.27% |  10.606 ms |       0.46% |   -113.268 us |  -1.06% |   FAIL   |
|   2^4   |     2^22      |  12.800 ms |       0.27% |  12.689 ms |       0.33% |   -110.459 us |  -0.86% |   FAIL   |
|   2^1   |     2^24      |  11.261 ms |      10.01% |  10.507 ms |       1.91% |   -754.056 us |  -6.70% |   FAIL   |
|   2^2   |     2^24      |  11.306 ms |      10.39% |  10.493 ms |       1.58% |   -812.631 us |  -7.19% |   FAIL   |
|   2^3   |     2^24      |  15.109 ms |       0.90% |  15.620 ms |      10.60% |    511.498 us |   3.39% |   FAIL   |
|   2^4   |     2^24      |  18.559 ms |       3.69% |  18.562 ms |       5.30% |      2.206 us |   0.01% |   PASS   |
|   2^1   |     2^26      |  30.345 ms |       4.05% |  29.219 ms |       4.78% |  -1126.209 us |  -3.71% |   PASS   |
|   2^2   |     2^26      |  30.467 ms |       4.22% |  29.164 ms |       4.85% |  -1303.629 us |  -4.28% |   FAIL   |
|   2^3   |     2^26      |  35.087 ms |       2.81% |  34.047 ms |       4.49% |  -1040.550 us |  -2.97% |   FAIL   |
|   2^4   |     2^26      |  44.072 ms |       3.11% |  41.409 ms |       3.47% |  -2662.539 us |  -6.04% |   FAIL   |
|   2^1   |     2^28      | 105.235 ms |       2.31% | 100.108 ms |       0.65% |  -5127.800 us |  -4.87% |   FAIL   |
|   2^2   |     2^28      | 105.108 ms |       1.54% | 100.133 ms |       0.50% |  -4975.085 us |  -4.73% |   FAIL   |
|   2^3   |     2^28      | 131.532 ms |       1.10% | 120.932 ms |       0.74% | -10599.756 us |  -8.06% |   FAIL   |
|   2^4   |     2^28      | 163.348 ms |       1.29% | 152.227 ms |       1.03% | -11120.952 us |  -6.81% |   FAIL   |
|   2^1   |     2^30      | 402.572 ms |       0.74% | 377.403 ms |       0.33% | -25169.766 us |  -6.25% |   FAIL   |
|   2^2   |     2^30      | 403.850 ms |       0.71% | 377.667 ms |       0.33% | -26182.859 us |  -6.48% |   FAIL   |
|   2^3   |     2^30      | 511.063 ms |       0.19% | 461.863 ms |       0.39% | -49200.109 us |  -9.63% |   FAIL   |
|   2^4   |     2^30      | 631.286 ms |       0.30% | 585.737 ms |       0.50% | -45548.761 us |  -7.22% |   FAIL   |

# json_read_data_type

## [0] Quadro GV100

|  data_type  |      io       |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|-------------|---------------|------------|-------------|------------|-------------|---------------|---------|----------|
|    FLOAT    | DEVICE_BUFFER | 769.726 ms |       0.21% | 710.914 ms |       0.24% | -58812.134 us |  -7.64% |   FAIL   |
|   DECIMAL   | DEVICE_BUFFER | 848.174 ms |       0.16% | 811.799 ms |       0.22% | -36374.756 us |  -4.29% |   FAIL   |
|   STRING    | DEVICE_BUFFER | 334.214 ms |       0.48% | 313.744 ms |       0.68% | -20469.883 us |  -6.12% |   FAIL   |
|    LIST     | DEVICE_BUFFER | 246.600 ms |       0.79% | 241.710 ms |       0.49% |  -4889.477 us |  -1.98% |   FAIL   |
|   STRUCT    | DEVICE_BUFFER | 936.482 ms |       0.25% | 854.513 ms |       0.48% | -81968.657 us |  -8.75% |   FAIL   |

# json_read_io

## [0] Quadro GV100

|      io       |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |         Diff |   %Diff |  Status  |
|---------------|------------|-------------|------------|-------------|--------------|---------|----------|
|   FILEPATH    | 549.848 ms |       1.97% | 549.053 ms |       2.05% |  -795.637 us |  -0.14% |   PASS   |
|  HOST_BUFFER  | 427.906 ms |       0.05% | 420.848 ms |       0.05% | -7058.240 us |  -1.65% |   FAIL   |
| DEVICE_BUFFER | 278.015 ms |       0.50% | 272.230 ms |       0.49% | -5784.574 us |  -2.08% |   FAIL   |

# Summary

- Total Matches: 43
  - Pass    (diff <= min_noise): 7
  - Unknown (infinite noise):    0
  - Failure (diff > min_noise):  36

File: cpp/src/io/json/json_tree.cu (hunk 4 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_tree.cu:631]: used the same probing used in distinct_count. What other probing options are there?

File: cpp/src/io/json/json_tree.cu (hunk 3 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_tree.cu:38]: replaced entired.

File: cpp/src/io/json/json_tree.cu [hunk 5]
```diff
@@ -632,23 +621,27 @@ std::pair<rmm::device_uvector<size_type>, rmm::device_uvector<size_type>> hash_n
...
+  auto nodes_itr          = thrust::make_counting_iterator<size_type>(0);
+  auto const num_inserted = key_set.insert(nodes_itr, nodes_itr + num_nodes, stream.value());
 
-  auto const num_columns = num_inserted;  // key_map.get_size() is not updated.
+  auto const num_columns = num_inserted;
   rmm::device_uvector<size_type> unique_keys(num_columns, stream);
-  key_map.retrieve_all(unique_keys.begin(), thrust::make_discard_iterator(), stream.value());
+  rmm::device_uvector<size_type> col_id(num_nodes, stream, mr);
+  key_set.find_async(nodes_itr, nodes_itr + num_nodes, col_id.begin(), stream.value());
+  std::ignore = key_set.retrieve_all(unique_keys.begin(), stream.value());
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply cpp/src/io/json/json_tree.cu:643]: Ideally, it should. We insert all nodes and retrieve all.

File: cpp/src/io/json/json_tree.cu (hunk 4 above)
Comment (ttnghia) [reply cpp/src/io/json/json_tree.cu:631]: @PointKernel?

Comment (PointKernel) [reply cpp/src/io/json/json_tree.cu:631]: We can switch between `linear_probing` and `double_hashing`. In general, `linear_probing` delivers better performance for set/map and `double_hashing` is preferred for multiset/multiset.

**Review by ttnghia — APPROVED** [review]

File: cpp/src/io/json/json_tree.cu (hunk 5 above)
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/io/json/json_tree.cu:642]: `retrieve_all` doesn't have the `_async` suffix which implies it synchronizes the given stream. We are using the same stream for `find_async` and `retrieve_all` thus relying on implicit sync inside `retrieve_all` is fine.

**Review by PointKernel — APPROVED** [review]
Looks great! Thanks!

**Pull request merged**